navapbc/labs-gen-ai-experiments
# Applying DSPy on Guru-based scenarios

## Ticket

Resolves https://navalabs.atlassian.net/browse/DST-168

## Changes

Use DSPy optimizers on prompt and question-answer examples provided in prompt.

## Context for reviewers

Results and findings are documented in https://navasage.atlassian.net/wiki/spaces/NL/pages/1072627716/Applying+DSPy+on+RAG+for+Household+scenarios

## Testing

Testing instructions and expected behavior:
1. `pip install -r requirements.txt`
1. Update `.env` (if needed)
1. Start openhermes LLM using Ollama: `ollama run openhermes` (Install Ollama from https://ollama.ai/)
2. Run `./dspy_engine.py`. For the question:
   - `press Enter to start from scratch` (a `compiled_module-optimizer3_10_openhermes_*.json` will be created)
   - choose `3` for 'BootstrapFewShot' optimizer
   - press Enter to choose `openhermes` for the LLM

Expected output:
```
...
----- optimizing ------------------
 60%|████████████████████████████████████████████████████████████████████████████████████████████████████▊                                                                   | 9/15 [01:51<01:14, 12.42s/it]
Bootstrapped 4 full traces after 10 examples in round 0.
--------------------------------------
(Elapsed time of optimize_module: 111.7682 seconds)
----- optimization ------------------
generate_answer
ChainOfThought(GenerateAnswer(context, question -> answer
    instructions='Given a question, promptly discern its context and components. Provide a direct response with either "Yes", "No", or "Maybe" based on the content and certainty of the query. Focus on clarity and decisiveness, being concise in conveying your evaluation while catering to the specifics of the question.\n'
    context = Field(annotation=str required=True json_schema_extra={'desc': 'may contain relevant facts used to answer the question', '__dspy_field_type': 'input', 'prefix': 'Context:'})
    question = Field(annotation=str required=True json_schema_extra={'__dspy_field_type': 'input', 'prefix': 'Question:', 'desc': '${question}'})
    answer = Field(annotation=str required=True json_schema_extra={'desc': 'Start with one of these words: Yes, No, Maybe', 'prefix': 'Response:', '__dspy_field_type': 'output'})
))
15 demos
...
--------------------------------------
----- evaluating against 15 items ------------------
Average Metric: 11.5 / 15  (76.7): 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 15/15 [03:03<00:00, 12.25s/it]
Average Metric: 11.5 / 15  (76.7%)
--------------------------------------
----- evalation results ------------------
1.0 | 1 | No | No, the client does
1.0 | 2 | Yes | Yes, the client must
0.5 | 3 | Maybe | No, the roommates do
1.0 | 4 | No | No, they cannot appl
0.5 | 5 | Maybe | Yes, the client can
0.5 | 6 | No | Based on the informa
1.0 | 7 | No | No, the client and h
1.0 | 8 | Yes | Yes, all family memb
1.0 | 9 | No | No, the client and h
0.5 | 10 | No | The 24 year old clie
1.0 | 11 | Yes | Yes, you are able to
0.5 | 12.1 | Yes | No, this child shoul
1.0 | 12.2 | Yes | Yes, the 23-year-old
0.5 | 12.3 | Maybe | The client has foste
0.5 | 12.4 | No | The client's mother

score: 76.67
8 (53%) correct
--------------------------------------
(Elapsed time of evaluate_module: 175.3017 seconds)
```

Unfortunately, evaluation accuracy varies between 33% and 53% for the same setup and parameters.
